ghcide/src/Development/IDE/Types/HscEnvEq.hs

```diff
@@ -112,10 +123,102 @@ newHscEnvEqWithImportPaths envImportPaths hscEnv deps = do
+            traverse_ (indexModuleHieFile hieDir) modIfaces
+        indexModuleHieFile :: FilePath -> ModIface -> IO ()
+        indexModuleHieFile hieDir modIface = do
+            let hiePath :: NormalizedFilePath
+                hiePath = toNormalizedFilePath' $
+                  hieDir </> toFilePath (moduleName $ mi_module modIface) ++ ".hie"
+            hieCheck <- checkHieFile recorder se "newHscEnvEqWithImportPaths" hiePath
+            case hieCheck of
+                HieFileMissing -> return ()
+                HieAlreadyIndexed -> return ()
+                CouldNotLoadHie _e -> return ()
+                DoIndexing hash hie ->
+                    indexHieFile se hiePath (FakeFile Nothing) hash hie
+        toFilePath :: ModuleName -> FilePath
+        toFilePath = separateDirectories . prettyModuleName
+            where
+                separateDirectories :: FilePath -> FilePath
+                separateDirectories moduleNameString =
+                    case breakOnDot moduleNameString of
+                        [] -> ""
+                        ms -> foldr1 (</>) ms
+                breakOnDot :: FilePath -> [FilePath]
+                breakOnDot = words . map replaceDotWithSpace
+                replaceDotWithSpace :: Char -> Char
+                replaceDotWithSpace '.' = ' '
+                replaceDotWithSpace c = c
+                prettyModuleName :: ModuleName -> String
+                prettyModuleName = filter (/= '"')
+                    . concat
+                    . drop 1
+                    . words
+                    . show
+        loadModIface :: Module -> IO (Maybe ModIface)
+        loadModIface m = do
+            modIface <- initIfaceLoad hscEnv $
+                loadInterface "" m (ImportByUser NotBoot)
+            return $ case modIface of
+                Maybes.Failed    _r -> Nothing
+                Maybes.Succeeded mi -> Just mi
+        packagesWithModules :: Map.Map Package [Module]
+        packagesWithModules = Map.fromSet getModulesForPackage packages
+        packages :: Set Package
+        packages = Set.fromList
+            $ map Package
+            $ Map.elems
+            $ getUnitInfoMap hscEnv
+        getModulesForPackage :: Package -> [Module]
+        getModulesForPackage (Package package) =
+            map makeModule allModules
+            where
+                allModules :: [(ModuleName, Maybe Module)]
+                allModules = unitExposedModules package
+                    ++ zip (unitHiddenModules package) (repeat Nothing)
+                makeModule :: (ModuleName, Maybe Module)
+                           -> Module
+                makeModule (moduleName, Nothing) =
+                    mkModule (unitInfoId package) moduleName
+                -- When module is re-exported from another package,
+                -- the origin module is represented by value in Just
+                makeModule (_, Just otherPackageMod) = otherPackageMod
```

Comment:
modules from other packages won't have a `.hie` file in the `extra-compilation` directory of the package under consideration, so we can just skip them from the list.